### PR TITLE
Fix granule duration check, to allow shorter than 48 scans in a granule

### PR DESCRIPTION
**Pull request by adybbroe**
Signed-off-by: Adam.Dybbroe <a000680@c21856.ad.smhi.se>

 - [ ] Closes #xxxx 
 - [x] Tests added 
 - [x] Tests passed: Passes ``pytest`` 
 - [ ] Passes ``flake8 nwcsafpps_runner`` 
 - [ ] Fully documented

### CODE REVIEW

**Comment by adybbroe** [comment]
@TAlonglong and @mraspaud Can we merge this one?

**Comment by mraspaud** [comment]
Not yet, tests are failing and it hasn't been reviewed :)

File: nwcsafpps_runner/tests/test_pps_hook.py [hunk 1]
```diff
@@ -231,7 +235,42 @@ def test_get_granule_duration(self, mandatory_param, filename):
...
+        sensor_is_viirs.return_value = True
+
+        metadata = self.metadata_with_start_and_end_times
+
+        posttroll_message = PostTrollMessage(0, metadata)
+        delta_t = timedelta(seconds=82.8)
+
+        with patch.object(PostTrollMessage, 'get_granule_duration', return_value=delta_t) as mock_method:
+            result = posttroll_message.is_segment()
+            self.assertTrue(result)
```
Comment (mraspaud) [inline nwcsafpps_runner/tests/test_pps_hook.py:266]: Could you also add a check for 48 scans?

**Review by mraspaud — CHANGES_REQUESTED** [review]
Thanks for fixing this. Could you make sure that 47 and 48 scans are the only possible values? (maybe add a reference to a document)

File: nwcsafpps_runner/tests/test_pps_hook.py (hunk 1 above)
Comment (adybbroe) [reply nwcsafpps_runner/tests/test_pps_hook.py:266]: I thought I had!?
I have a test for 82.2 seconds (47 scans) and 84.1 sec (48 scans).

The thing is, as far as I interpret things, that in the SDR files you have the start and end time in terms of time of the first scan, and time of the last scan. But to get to the duration of the entire granule you need to add the time length of one scan (~1.8 sec). Thus, the duration of 48 scans is ~85.9 sec and ~84.1 seconds for 47 scans.

**Comment by adybbroe** [comment]
> Not yet, tests are failing and it hasn't been reviewed :)

The couple of test-suites that fail (ubuntu/wondows latest), doesn't have anything to do with this PR as I see it.

**Comment by adybbroe** [comment]
> Thanks for fixing this. Could you make sure that 47 and 48 scans are the only possible values? (maybe add a reference to a document)

How do you mean? Should we omit PPS processing all together if there are 46 scans or less?

File: nwcsafpps_runner/tests/test_pps_hook.py (hunk 1 above)
Comment (mraspaud) [reply nwcsafpps_runner/tests/test_pps_hook.py:266]: I don't see the test for 84.1 seconds in `test_is_segment`, what am I missing?

**Comment by mraspaud** [comment]
> > Not yet, tests are failing and it hasn't been reviewed :)
> 
> The couple of test-suites that fail (ubuntu/wondows latest), doesn't have anything to do with this PR as I see it.

Could you just add a `skip` statement then for windows platforms if you don't want to fix it here?

**Comment by mraspaud** [comment]
> > Thanks for fixing this. Could you make sure that 47 and 48 scans are the only possible values? (maybe add a reference to a document)
> 
> How do you mean? Should we omit PPS processing all together if there are 46 scans or less?

I'm just unsure what the format specification says about viirs granules. Is is just 47 or 48 scans, or can it be less or more?

**Comment by adybbroe** [comment]
> > > Thanks for fixing this. Could you make sure that 47 and 48 scans are the only possible values? (maybe add a reference to a document)
> > 
> > 
> > How do you mean? Should we omit PPS processing all together if there are 46 scans or less?
> 
> I'm just unsure what the format specification says about viirs granules. Is is just 47 or 48 scans, or can it be less or more?

Yes, me too. From our own processing it looks like it is either 47 or 48 scans, but the actual length varies a bit more than just between two values. So, seems like 48 scans is not always exactly the same size in terms of start and end times in the files! I get betwee 1min24.1sec to 1min24.5 sec (that is actually just end-start times - so missing 1.779 sec) for the 48 scan files.

I have linked the SDR spec already, but don't find the allowed/occuring number of granules.

**Comment by adybbroe** [comment]
I have adjusted the granule length calculation adding length of one granule.

**Comment by mraspaud** [comment]
About skipping, I meant just skipping these specific tests, with https://docs.pytest.org/en/latest/skipping.html

**Comment by adybbroe** [comment]
> About skipping, I meant just skipping these specific tests, with https://docs.pytest.org/en/latest/skipping.html

Okay, thanks, but now I don't see what exactly was failing anymore, but it was something about gdal, and it was windows and ubuntu experimental, so I rather want to fix it in a separate PR then. The github ci was added just recently.
Ok?

File: nwcsafpps_runner/tests/test_pps_hook.py (hunk 1 above)
Comment (adybbroe) [reply nwcsafpps_runner/tests/test_pps_hook.py]: https://github.com/adybbroe/pytroll-pps-runner/blob/a0a7441893c4e5436474e009c9d9ea8858376ec3/nwcsafpps_runner/tests/test_pps_hook.py#L262

Comment (adybbroe) [reply nwcsafpps_runner/tests/test_pps_hook.py]: @mraspaud Do you see it now?

**Comment by adybbroe** [comment]
@mraspaud Where are we with this?

**Review by mraspaud — APPROVED** [review]
LGTM

**Pull request merged**